### PR TITLE
TypeLibrary

### PR DESCRIPTION
**Pull request by awcullen**
I've always admired how AspNetCore discovers all the controllers and views in the application.  I want this library to be able to discover the Encodable types from the referenced assemblies in the application.  

In this PR, I've introduced TypeLibrary and TypeLibraryAttribute. TypeLibraryAttribute is an attribute that marks assemblies that should be searched for encodable types. Example:

```
[assembly: Workstation.ServiceModel.Ua.TypeLibrary()]
namespace CustomTypeLibrary
{
    [DataTypeId("nsu=http://www.unifiedautomation.com/DemoServer/;i=3002")]
    [BinaryEncodingId("nsu=http://www.unifiedautomation.com/DemoServer/;i=5054")]
    public class CustomVector : Structure
    {
        public double X { get; set; }
        public double Y { get; set; }
        public double Z { get; set; }
    }
}

```

TypeLibrary is a singleton that searches the AppDomain for assemblies marked with  TypeLibraryAttribute and then searches the assemblies for types marked with BinaryEncodingIdAttribute.  The Types are stored in readonly dictionaries that the BinaryEncoder and BinaryDecoder access.

With this feature implemented, I was able to remove the AdditionalTypes parameter of the channel constructors.

### CODE REVIEW

**Comment by quinmars** [comment]
Without doubt, attributes are a convenient way to annotated types. But they are also inflexible. You cannot change or influence the annotation at runtime by code. That's why EF offers two ways to annotated types by attributes and by code. 
During a session there must be a bijective mapping between a C# class and a (binary) OPC UA encoding id, i.e., a C# type can have only one associated encoding id and vice versa. But currently (and this PR does not change that situation) you are forced to use the same type-encoding pair even between sessions. This would be ok in an ideal world. You or some one else would define a client/server interface (including the encoding ids) and this interface would be implemented on both sides. In reality, though, you would define a structural interface and take the encoding ids that the PLC server decides to use. Maybe one could use a proxy server to change the encoding ids, but that is extra work, extra costs and extra execution time. For a single HMI project this is not a problem, but @krzyfre already showed in #129 a situation were this is not solvable with opc-ua-client. Another thinkable situation would be a line controller, where you read and write product carrier tracking records from one to another machine in the line. If your machine suppliers happened to use different PLCs, you need to have the types defined for each PLC vendor and convert the object back and forth, although they are structurally equivalent and differ only in the encoding id. From a formal OPC UA point of view this might be correct, but it is tedious.

Second, the type library generates it's dictionary in a static constructor. I have thought about having a MEF plugin system in UaLayman where you can extend it to support special types, for example displaying an image. This is just a rough idea and maybe I will never implement it. Nonetheless, it's a use case that we should keep in mind. Due to the lazy static constructor execution in C#, this could still work, if the `TypeLibrary` is first accessed after the MEF composition. But it is brittle.

Maybe we can have the best of both worlds? An easy to use default behavior, and a way that allows more involved scenarios. If we combine our two PR, there could be a `TypeLibrary<BinaryEncodingAttribute>.CreateEncodingTable()` returning all type-encodingid pairs for the loaded assemblies. We could make this even the default behavior of the session constructor, if null is passed for the `additionalTypes` argument.

> Is it acceptable that type dictionaries are read-only and can not be added to after initialization?

As pointed out above, it depends what you understand on "after initialization". What I find interesting is that your `IEncodingContext` does not need any look up mechanism for the node id and type mapping. But I need some more time to dig into your changes.

**Comment by awcullen** [comment]
Thanks for looking though the PR.  In response to your first point about copying structures between two plcs that don't share the same EncodingId, please look at the following snippet where I show how to copy the ExtensionObject.  When you construct the second ExtensionObject, you are free to set a new EncodingId.

```
            var readRequest = new ReadRequest { 
                NodesToRead = new[] { 
                    new ReadValueId { 
                        NodeId = NodeId.Parse("ns=2;s=Demo.Static.Scalar.Vector"), 
                        AttributeId = AttributeIds.Value 
                    } 
                } 
            };
            var readResult = await channel.ReadAsync(readRequest);
            var v1 = (ExtensionObject)readResult.Results[0].Value;
            ExtensionObject v2;
            if (v1.BodyType == BodyType.Encodable)
            {
                // encoding id was found in type library and body was decoded.
                v2 = new ExtensionObject(
                        (IEncodable)v1.Body,
                        ExpandedNodeId.Parse("nsu=http://www.unifiedautomation.com/DemoServer/;i=5054")
                );
            }
            else
            {
                // encoding id was not found in type library and body is  byte array.
                v2 = new ExtensionObject(
                        (byte[])v1.Body,
                        ExpandedNodeId.Parse("nsu=http://www.unifiedautomation.com/DemoServer/;i=5054")
                );
            }
            var writeRequest = new WriteRequest
            {
                NodesToWrite = new[] {
                    new WriteValue {
                        NodeId = NodeId.Parse("ns=2;s=Demo.Static.Scalar.Vector"),
                        AttributeId = AttributeIds.Value,
                        Value = new DataValue(v2)
                    }
                }
            };
            var writeResponse = await channel.WriteAsync(writeRequest);
            StatusCode.IsGood(writeResponse.Results[0])
                .Should().BeTrue();

```

**Comment by quinmars** [comment]
> please look at the following snippet 

That is very interesting. But if I want to use the content of the vector I would need to decode it myself. Of course I could employ the existing `BinaryDecoder` but if the type contains futher structs I'd be lost. Because I cannot inject some other dictionary than the default one.

**Comment by awcullen** [comment]
Are there any other use cases for multiple TypeIds or multiple TypeLibraries?   
As in .NET, where values are not assignable to a variable of a different type, even if the fields are the same, if a Siemens and an Beckhoff plc both have variables that accept a Vector type, the TypeId should be the same.  The solution would be if PLC vendors begin allowing programmers to specify the TypeIds for the UDT.  Actually, I believe Siemens supports this today with a tool named SiOME https://support.industry.siemens.com/cs/attachments/109755133/109755133_SiOME_DOC_V21_en.pdf.

File: UaClient.UnitTests/IntegrationTests/IntegrationTests.cs [hunk 1]
```diff
@@ -41,13 +42,15 @@ public class IntegrationTests
...
         private readonly X509Identity x509Identity;
+        private readonly ITestOutputHelper output;
 
-        public IntegrationTests()
+        public IntegrationTests(ITestOutputHelper output)
         {
-            this.loggerFactory = LoggerFactory.Create(builder => builder.AddDebug());
-            this.logger = this.loggerFactory?.CreateLogger<IntegrationTests>();
+            this.output = output;
+            loggerFactory = LoggerFactory.Create(builder => builder.AddDebug());
```
[the remark this answers is not in the document]
Comment (awcullen) [reply UaClient.UnitTests/IntegrationTests/IntegrationTests.cs:50]: Good, I'll change the codebase to "_"

**Comment by quinmars** [comment]
> Actually, I believe Siemens supports this today with a tool named SiOME

Out of curiosity, I did yesterday some research and testing if Beckhoff's OPC UA server supports this. In some brochures it was mention that one can import a node set and map it to the PLC types. In a quick trail I wasn't capable to do so with their configurator, but it was the first time I used that tool. Sooner or later all PLC vendor need to support that, else you cannot implement a companion specification.

> Are there any other use cases for multiple TypeIds or multiple TypeLibraries?

No, not that I can think of. I thought about having several C# classes for one encoding id, for example to support different interface iteration steps. I haven't mentioned it in my first comment, because I do not think that this would be a good approach.

Multiple TypeLibraries are needed when we want to support encodings besides the binary encoding.

**Review by quinmars — APPROVED** [review]
Hi, Andrew. Athough I have in some places different ideas, I think, this PR is a clear improvement. I just want to ask you, to keep the duel targeting (netstandard2.0, metcoreapp3.0).

**Pull request merged**